libs/langchain/langchain/vectorstores/tiledb.py

```diff
@@ -0,0 +1,383 @@
+        )
+    return tiledb_vs
+
+
+class TileDB(VectorStore):
+    """Wrapper around TileDB vector database.
+
+    To use, you should have the ``tiledb-vector-search`` python package installed.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import TileDB
+            db = TileDB(embedding_function, index, metric, docs_array)
+
+    """
+
+    def __init__(
+        self,
+        embedding_function: Callable,
+        index: Any,
+        metric: str,
+        docs_array: tiledb.Array,
+    ):
+        """Initialize with necessary components."""
+        self.embedding_function = embedding_function
+        self.index = index
+        self.metric = metric
+        self.docs_array = docs_array
+
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        # TODO: Accept embedding object directly
+        return None
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        raise NotImplementedError(
+            "TileDB does not allow to add new data once the index is build."
+        )
+
+    def process_index_results(self, idxs: List[int]) -> List[Tuple[Document, float]]:
+        """Turns TileDB results into a list of documents and scores.
+
+        Args:
+            idxs: List of indices of the documents in the index.
+            dists: List of distances of the documents in the index.
+        Returns:
+            List of Documents and scores.
+        """
+        docs = []
+        for idx in idxs[0]:
+            doc = self.docs_array[idx]
+            if doc is None:
+                raise ValueError(f"Could not find document for id {idx}, got {doc}")
+            pickled_metadata = doc.get("metadata")
+            if pickled_metadata is not None:
+                metadata = pickle.loads(
+                    np.array(pickled_metadata.tolist()).astype(np.uint8).tobytes()
+                )
+                docs.append(
+                    (Document(page_content=str(doc["text"]), metadata=metadata), 1.0)
+                )
+            else:
+                docs.append(
+                    (Document(page_content=str(doc["text"])), 1.0)
+                )
+        return docs
+
+    def similarity_search_with_score_by_vector(
+        self, embedding: List[float], k: int = 4, search_k: int = -1
+    ) -> List[Tuple[Document, float]]:
+        """Return docs most similar to query.
+
+        Args:
+            query: Text to look up documents similar to.
+            k: Number of Documents to return. Defaults to 4.
+            search_k: inspect up to search_k nodes which defaults
+                to n_trees * n if not provided
+        Returns:
+            List of Documents most similar to the query and score for each
+        """
+        idxs = self.index.query(
+            np.array([np.array(embedding).astype(np.float32)]).astype(np.float32), k=k
```

Comment:
Is this cast guaranteed to be safe?

Reply:
(at least we should add a todo and a story to follow-up and clean it up later..)